### PR TITLE
docs: Add development guidelines for testing combinations

### PR DESCRIPTION
**Pull request by alexowens90**
Adds development guidelines section to the docs.

Initially, these just contain combinations of options and method parameters that should be considered when adding new features or fixing bugs, but could in future grow to encompass general coding guidelines.

### CODE REVIEW

**Review by poodlewars — APPROVED** [review]

File: docs/mkdocs/docs/technical/contributing.md [hunk 1]
```diff
@@ -332,3 +332,80 @@ To override that, use one of the following CMake variables*:
...
+Due to ArcticDB being a purely client-side database, it is possible for data to be written by clients that are of later versions than clients that need to read the same data.
+If the change is breaking, such that older clients will not be able to read the data, then this should be clearly documented both in the PR and in the online documentation.
+If possible, a version number should also be added such that future changes in the same area will display a more clear error message.1
+
+### API
+
+Any changes to the API (including when exceptions are thrown and the type of the exception) must be weighed up against the change breaking behaviour for existing users.
+This is particularly true of the `NativeVersionStore` API, as this has many users inside Man Group.
+
+### Snapshots
```
Comment (jjerphan) [inline docs/mkdocs/docs/technical/contributing.md:356]: What is a snapshot?

File: docs/mkdocs/docs/technical/contributing.md [hunk 2]
```diff
@@ -332,3 +332,80 @@ To override that, use one of the following CMake variables*:
 
    (Only) `Python_ROOT_DIR` can also be set as an environment variable.
    Setting the others in the environment might have no effect.)
+
+Development Guidelines
+======================
+
+ArcticDB has a lot of write-time options to configure how the data is stored on disk.
+When adding new features, it is important to test not only the simplest case, but also how the new feature interacts with these various options.
+This section serves as a guide as to what should be considered when adding a new feature.1
```
Comment (vasil-pashov) [inline docs/mkdocs/docs/technical/contributing.md:341]:
```suggestion
This section serves as a guide as to what should be considered when adding a new feature.
```

File: docs/mkdocs/docs/technical/contributing.md [hunk 3]
```diff
@@ -332,3 +332,80 @@ To override that, use one of the following CMake variables*:
...
+
+## Backwards compatibility
+
+### Data stored on disk
+
+Due to ArcticDB being a purely client-side database, it is possible for data to be written by clients that are of later versions than clients that need to read the same data.
+If the change is breaking, such that older clients will not be able to read the data, then this should be clearly documented both in the PR and in the online documentation.
+If possible, a version number should also be added such that future changes in the same area will display a more clear error message.1
```
Comment (vasil-pashov) [inline docs/mkdocs/docs/technical/contributing.md:349]:
```suggestion
If possible, a version number should also be added such that future changes in the same area will display a more clear error message.
```

**Review by jjerphan — COMMENTED** [review]
Thank you for adding those pieces of documentation, @alexowens90.

File: docs/mkdocs/docs/technical/contributing.md [hunk 4]
```diff
@@ -332,3 +332,80 @@ To override that, use one of the following CMake variables*:
...
+
+### Data stored on disk
+
+Due to ArcticDB being a purely client-side database, it is possible for data to be written by clients that are of later versions than clients that need to read the same data.
+If the change is breaking, such that older clients will not be able to read the data, then this should be clearly documented both in the PR and in the online documentation.
+If possible, a version number should also be added such that future changes in the same area will display a more clear error message.
+
+### API
+
+Any changes to the API (including when exceptions are thrown and the type of the exception) must be weighed up against the change breaking behaviour for existing users.
```
Comment (mehertz) [inline docs/mkdocs/docs/technical/contributing.md:353]:
```suggestion
Any changes to the API (including when exceptions are thrown and the type of the exception) must be weighed up against the change breaking behaviour for existing users.
Please make this as clear to reviewers as possible by ensuring **API changes are clearly described in the PR description**. 
If the change is breaking, please also ensure that that is appropriately highlighted in the PR description as well.
```

**Review by mehertz — CHANGES_REQUESTED** [review]

File: docs/mkdocs/docs/technical/contributing.md (hunk 1 above)
Comment (mehertz) [reply docs/mkdocs/docs/technical/contributing.md:356]: https://docs.arcticdb.io/tutorials/snapshots/

https://docs.arcticdb.io/api/library#arcticdb.version_store.library.Library.snapshot

Comment (mehertz) [reply docs/mkdocs/docs/technical/contributing.md:356]: Might be worth linking here actually.

File: docs/mkdocs/docs/technical/contributing.md [hunk 5]
```diff
@@ -332,3 +332,80 @@ To override that, use one of the following CMake variables*:
...
+
+The majority of data written into ArcticDB uses dense columns, meaning that within a given row-slice, every row has an associated value in every column (even if that value is `NaN`).
+The concept of sparse columns is implemented as well, where columns may have values missing for none, some, or all rows within a row-slice.
+Any read-like method should be able to handle both types of stored column.
+
+## Pickling
+
+When data that cannot be normalized to a supported data type in ArcticDB, it can still be stored using `write_pickle` and similar.
+There are many operations that cannot be performed on pickled data, such as `append`, `update`, `date_range` search, and many more.
+It is important that if a user attempts an operation that is not supported with pickled data, that they receive a helpful error message.
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply docs/mkdocs/docs/technical/contributing.md:413]: In the `Library` API it is a separate method called `write_pickle` so people cannot do it by accident.

**Review by mehertz — APPROVED** [review]

**Pull request merged**